cmd/govim/fillstruct.go

```diff
@@ -34,33 +35,112 @@ func (v *vimstate) fillStruct(flags govim.CommandFlags, args ...string) error {
+	// version of the document.
+
+	// The gopls ExecuteCommand is blocking, and gopls will call back to govim
+	// using ApplyEdit that must be handled before the blocking is released.
+	// Since fillstruct is ordered by the user (and the single threaded nature
+	// of vim), we are effectively blocking ApplyEdit from modifying buffers.
+	//
+	// To prevent a deadlock, we create a channel that ApplyEdit can use pass
+	// edits to this thread (if needed). And then call ExecuteCommand in a
+	// separate goroutine so that this thread can go on updating buffers
+	// until the ExecuteCommand is released. When it is, we implicitly know
+	// that ApplyEdit has been processed.
+	editsCh := make(chan applyEditCall)
+	v.govimplugin.applyEditsLock.Lock()
+	v.govimplugin.applyEditsCh = editsCh
+	v.govimplugin.applyEditsLock.Unlock()
+	done := make(chan struct{})
+
+	v.tomb.Go(func() error {
+		// We can only apply one command at the moment since they all target the same document
+		// version. Let's go for the first one and let the user call fillstruct again if they
+		// want to fill several structs on the same line.
+		ca := codeActions[0]
+		_, err := v.server.ExecuteCommand(context.Background(), &protocol.ExecuteCommandParams{
+			Command:                ca.Command.Command,
+			Arguments:              ca.Command.Arguments,
+			WorkDoneProgressParams: protocol.WorkDoneProgressParams{},
+		})
+		if err != nil {
+			v.Logf("executeCommand failed: %v", err)
+		}
+
+		v.govimplugin.applyEditsLock.Lock()
+		v.govimplugin.applyEditsCh = nil
+		v.govimplugin.applyEditsLock.Unlock()
+		close(done)
+		return nil
+	})
+
+	for {
+		select {
+		case <-done:
+			return nil
+		case c := <-editsCh:
+			res, err := v.applyWorkspaceEdit(c.params)
+			c.responseCh <- applyEditResponse{res, err}
+		}
+	}
+}
+
+// applyEditCall represents a single LSP ApplyEdit call including a channel used
+// to pass a response back.
+type applyEditCall struct {
+	params     *protocol.ApplyWorkspaceEditParams
+	responseCh chan applyEditResponse
+}
+
+// applyEditResponse represents a LSP ApplyEdit response
+type applyEditResponse struct {
+	res *protocol.ApplyWorkspaceEditResponse
+	err error
+}
+
+func (v *vimstate) applyWorkspaceEdit(params *protocol.ApplyWorkspaceEditParams) (*protocol.ApplyWorkspaceEditResponse, error) {
+	res := &protocol.ApplyWorkspaceEditResponse{Applied: true}
+
+	edits := make(map[*types.Buffer][]protocol.TextEdit)
+	for _, dc := range params.Edit.DocumentChanges {
+		// verify that the version of the edits matches a buffer
+		var buf *types.Buffer
+		for _, b := range v.buffers {
+			if b.URI() != span.URI(dc.TextDocument.URI) {
+				continue
 			}
+
 			if ev := int(math.Round(dc.TextDocument.Version)); ev > 0 && ev != b.Version {
-				return fmt.Errorf("got edits for version %v, but current buffer version is %v", ev, b.Version)
+				v.Logf("got edits for buffer version %v, found matching buffer with version %v", ev, b.Version)
```

Comment:
This feels like it should be an error to me. i.e. we've managed to get ourselves into a state where we were given an edit for the wrong version of a document. There might be conditions under which "best efforts" is acceptable, but I'd want those to be very clearly identified and commented.

Reply:
Good point, was going for "best efforts" here but an error is better.